core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1865,6 +1876,39 @@ abstract class RDD[T: ClassTag](
   // RDD chain.
   @transient protected lazy val isBarrier_ : Boolean =
     dependencies.filter(!_.isInstanceOf[ShuffleDependency[_, _, _]]).exists(_.rdd.isBarrier())
+
+  /**
+   * Returns the random level of this RDD's computing function. Please refer to [[RDD.RandomLevel]]
+   * for the definition of random level.
+   *
+   * By default, an RDD without parents(root RDD) is IDEMPOTENT. For RDDs with parents, the random
+   * level of current RDD is the random level of the parent which is random most.
+   */
+  // TODO: make it public so users can set random level to their custom RDDs.
+  // TODO: this can be per-partition. e.g. UnionRDD can have different random level for different
+  // partitions.
+  private[spark] def computingRandomLevel: RDD.RandomLevel.Value = {
+    val parentRandomLevels = dependencies.map {
+      case dep: ShuffleDependency[_, _, _] =>
+        if (dep.rdd.computingRandomLevel == RDD.RandomLevel.INDETERMINATE) {
+          RDD.RandomLevel.INDETERMINATE
```

Comment:
It does not matter what the parent RDD's order was - after shuffle, currently, it is going to be always UNORDERED - unless Aggregator and key ordering is specified in dep.
Given [this comment](https://github.com/apache/spark/pull/22112#issuecomment-414034703), and adding a few missing cases, this becomes:
```
  // If checkpointed already - then always same order
  case dep: Dependency if dep.rdd.isCheckpointed => RDD.RandomLevel.IDEMPOTENT
  // if same partitioner, then shuffle not done.
  case dep: ShuffleDependency[_, _, _] if dep.partitioner == partitioner => dep.rdd.computingRandomLevel
  // if aggregator specified (and so unique keys) and key ordering specified - then consistent ordering.
  case dep: ShuffleDependency[_, _, _] if dep.keyOrdering.isDefined && dep.aggregator.isDefined =>
    RDD.RandomLevel.IDEMPOTENT
  // All other shuffle cases, we dont know the output order in spark.
  case dep: ShuffleDependency[_, _, _] => RDD.RandomLevel.INDETERMINATE
```

Reply:
> If checkpointed already - then always same order

`RDD#dependencies` is defined as
```
final def dependencies: Seq[Dependency[_]] = {
    checkpointRDD.map(r => List(new OneToOneDependency(r))).getOrElse {
      if (dependencies_ == null) {
        dependencies_ = getDependencies
      }
      dependencies_
    }
  }
```
So we don't need to handle checkpoint here.

> All other shuffle cases, we dont know the output order in spark.

Actually we know. As long as the shuffle map stage RDD is IDEMPOTENT or UNORDERED, the reduce RDD is UNORDERED instead of INDETERMINATE.

Reply:
> > All other shuffle cases, we dont know the output order in spark.
> 
> Actually we know. As long as the shuffle map stage RDD is IDEMPOTENT or UNORDERED, the reduce RDD is UNORDERED instead of INDETERMINATE.

IIUC shuffle map itself works as follows:

- If Aggregator and key ordering are specified:
  - output becomes idempotent;
- If Aggregator or key ordering are not specified:
  - If input is indeterminate, then output becomes indeterminate;
  - If input is idempotent or unordered, then output becomes unordered.

We have to also include the case @mridulm raised that shuffle map may be skipped.

Reply:
RE: checkpoint.

I wanted to handle two cases.
* Checkpoint is being done as part of the current job (and not a previous job which forced materialization of checkpoint'ed RDD).
* Checkpoint is happening to reliable store, not local - where we are subject to failures on node failures.

Looks like `dep.rdd.isCheckpointed` is the wrong way to go about it (relying on `dependencies` is insufficient for both cases).

A better option seems to be:
```
  // If checkpointed already - then always same order
  case dep: Dependency if dep.rdd.getCheckpointFile.isDefined => RDD.RandomLevel.IDEMPOTENT
```

> Actually we know. As long as the shuffle map stage RDD is IDEMPOTENT or UNORDERED, the reduce RDD is UNORDERED instead of INDETERMINATE.

It does not matter what the output order of map stage was, after we shuffle the map output, it is always indeterminate order except for the specific cases I referred to above.

Reply:
Crap, brain fart ... you are right it is UNORDERED and not INDETERMINATE ... I am still getting my head around the terms :-(